In [1]:
import graphlab
graphlab.canvas.set_target('browser')

# Load the CIFAR-10 dataset

In [2]:
image_train = graphlab.SFrame('image_train_data/')

This non-commercial license of GraphLab Create for academic use is assigned to mohammed303760@science.tanta.edu.eg and will expire on February 03, 2018.


[INFO] graphlab.cython.cy_server: GraphLab Create v2.1 started. Logging: /tmp/graphlab_server_1489158386.log


In [3]:
image_train.head()

Columns:
	id	int
	image	Image
	label	str
	deep_features	array
	image_array	array

Rows: 10

Data:
+-----+----------------------+------------+-------------------------------+
|  id |        image         |   label    |         deep_features         |
+-----+----------------------+------------+-------------------------------+
|  24 | Height: 32 Width: 32 |    bird    | [0.242871761322, 1.0954537... |
|  33 | Height: 32 Width: 32 |    cat     | [0.525087952614, 0.0, 0.0,... |
|  36 | Height: 32 Width: 32 |    cat     | [0.566015958786, 0.0, 0.0,... |
|  70 | Height: 32 Width: 32 |    dog     | [1.12979578972, 0.0, 0.0, ... |
|  90 | Height: 32 Width: 32 |    bird    | [1.71786928177, 0.0, 0.0, ... |
|  97 | Height: 32 Width: 32 | automobile | [1.57818555832, 0.0, 0.0, ... |
| 107 | Height: 32 Width: 32 |    dog     | [0.0, 0.0, 0.220677852631,... |
| 121 | Height: 32 Width: 32 |    bird    | [0.0, 0.23753464222, 0.0, ... |
| 136 | Height: 32 Width: 32 | automobile | [0.0, 0.0, 0.0, 0.0, 0.0, ... |
| 138 | Height: 32 Width: 32 |    bird    | [0.658935725689, 0.0, 0.0,... |
+-----+----------------------+------------+-------------------------------+
+-------------------------------+
|          image_array          |
+-------------------------------+
| [73.0, 77.0, 58.0, 71.0, 6... |
| [7.0, 5.0, 8.0, 7.0, 5.0, ... |
| [169.0, 122.0, 65.0, 131.0... |
| [154.0, 179.0, 152.0, 159.... |
| [216.0, 195.0, 180.0, 201.... |
| [33.0, 44.0, 27.0, 29.0, 4... |
| [97.0, 51.0, 31.0, 104.0, ... |
| [93.0, 96.0, 88.0, 102.0, ... |
| [35.0, 59.0, 53.0, 36.0, 5... |
| [205.0, 193.0, 195.0, 200.... |
+-------------------------------+
[10 rows x 5 columns]

# Train a nearest-neighbors model for retrieving images using deep features

In [4]:
knn_model = graphlab.nearest_neighbors.create(image_train,
                                             features=['deep_features'],
                                             label='id')

Starting brute force nearest neighbors model training.

# Use image retieval model with deep features to find similar images

In [5]:
cat = image_train[18:19]
cat['image'].show()

Canvas is accessible via web browser at the URL: http://localhost:36783/index.html
Opening Canvas in default web browser.


In [6]:
knn_model.query(cat)

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.0498753   | 69.706ms     |

| Done         |         | 100         | 331.446ms    |

+--------------+---------+-------------+--------------+

Columns:
	query_label	int
	reference_label	int
	distance	float
	rank	int

Rows: 5

Data:
+-------------+-----------------+---------------+------+
| query_label | reference_label |    distance   | rank |
+-------------+-----------------+---------------+------+
|      0      |       384       |      0.0      |  1   |
|      0      |       6910      | 36.9403137951 |  2   |
|      0      |      39777      | 38.4634888975 |  3   |
|      0      |      36870      | 39.7559623119 |  4   |
|      0      |      41734      | 39.7866014148 |  5   |
+-------------+-----------------+---------------+------+
[5 rows x 4 columns]

In [8]:
def get_images_from_ids(query_result):
    return image_train.filter_by(query_result['reference_label'], 'id')

In [9]:
cat_neighbors = get_images_from_ids(knn_model.query(cat))

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.0498753   | 13.598ms     |

| Done         |         | 100         | 334.746ms    |

+--------------+---------+-------------+--------------+

In [10]:
cat_neighbors['image'].show()

Canvas is accessible via web browser at the URL: http://localhost:36783/index.html
Opening Canvas in default web browser.


# Quering for the most similar images for car image

In [22]:
car = image_train[8:9]
car['image'].show()

Canvas is accessible via web browser at the URL: http://localhost:36783/index.html
Opening Canvas in default web browser.


In [23]:
get_images_from_ids(knn_model.query(car))['image'].show()

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.0498753   | 17.523ms     |

| Done         |         | 100         | 318.469ms    |

+--------------+---------+-------------+--------------+

Canvas is accessible via web browser at the URL: http://localhost:36783/index.html
Opening Canvas in default web browser.


# Displaying other example image retrievals with Python lambda

# Just for fun, let's create a lambda to find and show nearest neighbor images

In [24]:
show_neighbors = lambda i: get_images_from_ids(knn_model.query(image_train[i:i+1]))['image'].show()

In [25]:
show_neighbors(8)

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.0498753   | 16.019ms     |

| Done         |         | 100         | 315.46ms     |

+--------------+---------+-------------+--------------+

Canvas is accessible via web browser at the URL: http://localhost:36783/index.html
Opening Canvas in default web browser.


In [26]:
show_neighbors(26)

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.0498753   | 21.254ms     |

| Done         |         | 100         | 334.727ms    |

+--------------+---------+-------------+--------------+

Canvas is accessible via web browser at the URL: http://localhost:36783/index.html
Opening Canvas in default web browser.


In [27]:
show_neighbors(1222)

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.0498753   | 18.219ms     |

| Done         |         | 100         | 330.423ms    |

+--------------+---------+-------------+--------------+

Canvas is accessible via web browser at the URL: http://localhost:36783/index.html
Opening Canvas in default web browser.


In [28]:
show_neighbors(2000)

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.0498753   | 15.108ms     |

| Done         |         | 100         | 326.569ms    |

+--------------+---------+-------------+--------------+

Canvas is accessible via web browser at the URL: http://localhost:36783/index.html
Opening Canvas in default web browser.


# Quiz

# Quiz (1)
## What’s the least common category in the training data?
* bird (Selected)
* dog
* cat
* automobile

## Computing summary statistics of the data: Sketch summaries are techniques for computing summary statistics of data very quickly. In GraphLab Create, SFrames and SArrays include a method:
* .sketch_summary()
### which computes such summary statistics. Using the training data, compute the sketch summary of the ‘label’ column and interpret the results. What’s the least common category in the training data?

In [41]:
quiz_image_train = graphlab.SFrame('image_train_data/')
quiz_image_train['label'].sketch_summary()


+------------------+-------+----------+
|       item       | value | is exact |
+------------------+-------+----------+
|      Length      |  2005 |   Yes    |
| # Missing Values |   0   |   Yes    |
| # unique values  |   4   |    No    |
+------------------+-------+----------+

Most frequent items:
+-------+------------+-----+-----+------+
| value | automobile | cat | dog | bird |
+-------+------------+-----+-----+------+
| count |    509     | 509 | 509 | 478  |
+-------+------------+-----+-----+------+


# Quiz (2)
## Of the images below, which is the nearest ‘cat’ labeled image in the training data to the the first image in the test data (image_test[0:1])?

<img src='Images/Question_2/1.png'>

<img src='Images/Question_2/2.png'>

<img src='Images/Question_2/3.png'>

<img src='Images/Question_2/4.png'>

<img src='Images/Question_2/5.png'>

* Selected
<img src='Images/Question_2/6.png'>

## 2. Creating category-specific image retrieval models: In most retrieval tasks, the data we have is unlabeled, thus we call these unsupervised learning problems. However, we have labels in this image dataset, and will use these to create one model for each of the 4 image categories, {‘dog’,’cat’,’automobile’,bird’}. To start, follow these steps:

* Split the SFrame with the training data into 4 different SFrames. Each of these will contain data for 1 of the 4 categories above. Hint: if you use a logical filter to select the rows where the ‘label’ column equals ‘dog’, you can create an SFrame with only the data for images labeled ‘dog’.
* Similarly to the image retrieval notebook you downloaded, you are going to create a nearest neighbor model using the 'deep_features' as the features, but this time create one such model for each category, using the corresponding subset of the training_data. You can call the model with the ‘dog’ data the dog_model, the one with the ‘cat’ data the cat_model, as so on.

In [46]:
dog_image_train = quiz_image_train[quiz_image_train['label'] == 'dog']
print len(dog_image_train)
cat_image_train = quiz_image_train[quiz_image_train['label'] == 'cat']
print len(cat_image_train)
automobile_image_train = quiz_image_train[quiz_image_train['label'] == 'automobile']
print len(automobile_image_train)
bird_image_train = quiz_image_train[quiz_image_train['label'] == 'bird']
print len(bird_image_train)

509
509
509
478


In [48]:
dog_model = graphlab.nearest_neighbors.create(dog_image_train,
                                             features=['deep_features'],
                                             label='id')
cat_model = graphlab.nearest_neighbors.create(cat_image_train,
                                             features=['deep_features'],
                                             label='id')
automobile_model = graphlab.nearest_neighbors.create(automobile_image_train,
                                                    features=['deep_features'],
                                                    label='id')
bird_model = graphlab.nearest_neighbors.create(bird_image_train,
                                              features=['deep_features'],
                                              label='id')

Starting brute force nearest neighbors model training.

Starting brute force nearest neighbors model training.

Starting brute force nearest neighbors model training.

Starting brute force nearest neighbors model training.

### Using these models, answer the following questions. The cat image below is the first in the test data:

In [50]:
quiz_test_data = graphlab.SFrame('image_test_data/')
cat = quiz_test_data[0:1]
cat['image'].show()

Canvas is updated and available in a tab in the default browser.


In [51]:
def get_images(query_result):
    return quiz_image_train.filter_by(query_result['reference_label'], 'id')

In [52]:
nearest_cats = get_images(cat_model.query(cat))

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.196464    | 112.778ms    |

| Done         |         | 100         | 191.838ms    |

+--------------+---------+-------------+--------------+

In [53]:
nearest_cats['image'].show()

Canvas is updated and available in a tab in the default browser.


# Quiz (3)
## Of the images below, which is the nearest ‘dog’ labeled image in the training data to the the first image in the test data (image_test[0:1])?

<img src='Images/Question_3/1.png'>

<img src='Images/Question_3/2.png'>

<img src='Images/Question_3/3.png'>

* Selected
<img src='Images/Question_3/4.png'>

<img src='Images/Question_3/5.png'>

<img src='Images/Question_3/6.png'>

In [54]:
nearest_dogs = get_images(dog_model.query(cat))

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.196464    | 25.837ms     |

| Done         |         | 100         | 114.124ms    |

+--------------+---------+-------------+--------------+

In [56]:
nearest_dogs['image'].show()

Canvas is updated and available in a tab in the default browser.


# Quiz (4)
## For the first image in the test data, in what range is the mean distance between this image and its 5 nearest neighbors that were labeled ‘cat’ in the training data?
* 33 to 35
* 35 to 37 (Selected)
* 37 to 39
* 39 to 41
* Above 41

# Quiz (5)
## For the first image in the test data, in what range is the mean distance between this image and its 5 nearest neighbors that were labeled ‘dog’ in the training data?
* 33 to 35
* 35 to 37
* 37 to 39 (Selected)
* 39 to 41
* Above 41

# Quiz (6)
## On average, is the first image in the test data closer to its 5 nearest neighbors in the ‘cat’ data or in the ‘dog’ data?
* cat (Selected)
* dog

### 3. A simple example of nearest-neighbors classification: When we queried a nearest neighbors model, the ‘distance’ column in the table above shows the computed distance between the input and each of the retrieved neighbors. In this question, you will use these distances to perform a classification task, using the idea of a nearest-neighbors classifier.
* For the first image in the test data (image_test[0:1]), which we used above, compute the mean distance between this image at its 5 nearest neighbors that were labeled ‘cat’ in the training data (similarly to what you did in the previous question). Save this result.
* Similarly, for the first image in the test data (image_test[0:1]), which we used above, compute the mean distance between this image at its 5 nearest neighbors that were labeled ‘dog’ in the training data (similarly to what you did in the previous question). Save this result.
* On average, is the first image in the test data closer to its 5 nearest neighbors in the ‘cat’ data or in the ‘dog’ data? (In a later course, we will see that this is an example of what is called a k-nearest neighbors classifier, where we use the label of neighboring points to predict the label of a test point.)

In [58]:
cat_model.query(cat)

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.196464    | 40.864ms     |

| Done         |         | 100         | 164.684ms    |

+--------------+---------+-------------+--------------+

Columns:
	query_label	int
	reference_label	int
	distance	float
	rank	int

Rows: 5

Data:
+-------------+-----------------+---------------+------+
| query_label | reference_label |    distance   | rank |
+-------------+-----------------+---------------+------+
|      0      |      16289      |  34.623719208 |  1   |
|      0      |      45646      | 36.0068799284 |  2   |
|      0      |      32139      | 36.5200813436 |  3   |
|      0      |      25713      | 36.7548502521 |  4   |
|      0      |       331       | 36.8731228168 |  5   |
+-------------+-----------------+---------------+------+
[5 rows x 4 columns]

In [61]:
cat_model.query(cat)['distance'].mean()

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.196464    | 28.306ms     |

| Done         |         | 100         | 113.137ms    |

+--------------+---------+-------------+--------------+

36.15573070978294

In [59]:
dog_model.query(cat)

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.196464    | 42.658ms     |

| Done         |         | 100         | 139.839ms    |

+--------------+---------+-------------+--------------+

Columns:
	query_label	int
	reference_label	int
	distance	float
	rank	int

Rows: 5

Data:
+-------------+-----------------+---------------+------+
| query_label | reference_label |    distance   | rank |
+-------------+-----------------+---------------+------+
|      0      |      16976      | 37.4642628784 |  1   |
|      0      |      13387      | 37.5666832169 |  2   |
|      0      |      35867      | 37.6047267079 |  3   |
|      0      |      44603      | 37.7065585153 |  4   |
|      0      |       6094      | 38.5113254907 |  5   |
+-------------+-----------------+---------------+------+
[5 rows x 4 columns]

In [62]:
dog_model.query(cat)['distance'].mean()

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.196464    | 35.122ms     |

| Done         |         | 100         | 139.69ms     |

+--------------+---------+-------------+--------------+

37.77071136184156

# Quiz (7)
## In what range is the accuracy of the 1-nearest neighbor classifier at classifying ‘dog’ images from the test set?
* 50 to 60
* 60 to 70 (Selected)
* 70 to 80
* 80 to 90
* 90 to 100

## 4. [Challenging Question] Computing nearest neighbors accuracy using SFrame operations: A nearest neighbor classifier predicts the label of a point as the most common label of its nearest neighbors. In this question, we will measure the accuracy of a 1-nearest-neighbor classifier, i.e., predict the output as the label of the nearest neighbor in the training data. Although there are simpler ways of computing this result, we will go step-by-step here to introduce you to more concepts in nearest neighbors and SFrames, which will be useful later in this Specialization.
* Training models: For this question, you will need the nearest neighbors models you learned above on the training data, i.e., the dog_model, cat_model, automobile_model and bird_model.
* Spliting test data by label: Above, you split the train data SFrame into one SFrame for images labeled ‘dog’, another for those labeled ‘cat’, etc. Now, do the same for the test data. You can call the resulting SFrames

In [65]:
# quiz_image_train - quiz_test_data
# cat_image_train - dog_image_train - automobile_image_train - bird_image_train
# cat_model - dog_model - automobile_model - bird_model
# cat_image_test - dog_image_test - automobile_image_test - bird_image_test

In [69]:
cat_image_test = quiz_test_data[quiz_test_data['label'] == 'cat']
dog_image_test = quiz_test_data[quiz_test_data['label'] == 'dog']
automobile_image_test = quiz_test_data[quiz_test_data['label'] == 'automobile']
bird_image_test = quiz_test_data[quiz_test_data['label'] == 'bird']

### Finding nearest neighbors in the training set for each part of the test set: Thus far, we have queried, e.g.,
* our nearest neighbors models with a single image as the input, but you can actually query with a whole set of data, and it will find the nearest neighbors for each data point. Note that the input index will be stored in the ‘query_label’ column of the output SFrame.
* Using this knowledge find the closest neighbor in to the dog test data using each of the trained models, e.g.,
    * dog_cat_neighbors = cat_model.query(dog_image_test, k=1)
* finds 1 neighbor (that’s what k=1 does) to the dog test images (image_test_dog) in the cat portion of the training data (used to train the cat_model). 

In [71]:
dog_cat_neighbors = cat_model.query(dog_image_test, k=1)

Starting blockwise querying.

max rows per data block: 4348

number of reference data blocks: 4

number of query data blocks: 1

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 1000         | 127000  | 24.9509     | 802.027ms    |

| Done         | 509000  | 100         | 859.829ms    |

+--------------+---------+-------------+--------------+

### Now, do this for every combination of the labels in the training and test data.

In [72]:
dog_cat_neighbors = cat_model.query(dog_image_test, k=1)
dog_dog_neighbors = dog_model.query(dog_image_test, k=1)
dog_bird_neighbors = bird_model.query(dog_image_test, k=1)
dog_automobile_neighbors = automobile_model.query(dog_image_test, k=1)
cat_cat_neighbors = cat_model.query(cat_image_test, k=1)
cat_dog_neighbors = dog_model.query(cat_image_test, k=1)
cat_bird_neighbors = bird_model.query(cat_image_test, k=1)
cat_automobile_neighbors = automobile_model.query(cat_image_test, k=1)
bird_cat_neighbors = cat_model.query(bird_image_test, k=1)
bird_dog_neighbors = dog_model.query(bird_image_test, k=1)
bird_bird_neighbors = bird_model.query(bird_image_test, k=1)
bird_automobile_neighbors = automobile_model.query(bird_image_test, k=1)
automobile_cat_neighbors = cat_model.query(automobile_image_test, k=1)
automobile_dog_neighbors = dog_model.query(automobile_image_test, k=1)
automobile_bird_neighbors = bird_model.query(automobile_image_test, k=1)
automobile_automobile_neighbors = automobile_model.query(automobile_image_test, k=1)

Starting blockwise querying.

max rows per data block: 4348

number of reference data blocks: 4

number of query data blocks: 1

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 1000         | 127000  | 24.9509     | 557.006ms    |

| Done         | 509000  | 100         | 627.239ms    |

+--------------+---------+-------------+--------------+

Starting blockwise querying.

max rows per data block: 4348

number of reference data blocks: 4

number of query data blocks: 1

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 1000         | 128000  | 25.1473     | 512.269ms    |

| Done         | 509000  | 100         | 608.746ms    |

+--------------+---------+-------------+--------------+

Starting blockwise querying.

max rows per data block: 4348

number of reference data blocks: 4

number of query data blocks: 1

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 1000         | 119000  | 24.8954     | 503.885ms    |

| Done         | 478000  | 100         | 556.191ms    |

+--------------+---------+-------------+--------------+

Starting blockwise querying.

max rows per data block: 4348

number of reference data blocks: 4

number of query data blocks: 1

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 1000         | 127000  | 24.9509     | 501.122ms    |

| Done         | 509000  | 100         | 680.769ms    |

+--------------+---------+-------------+--------------+

Starting blockwise querying.

max rows per data block: 4348

number of reference data blocks: 4

number of query data blocks: 1

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 1000         | 127000  | 24.9509     | 517.204ms    |

| Done         | 509000  | 100         | 565.784ms    |

+--------------+---------+-------------+--------------+

Starting blockwise querying.

max rows per data block: 4348

number of reference data blocks: 4

number of query data blocks: 1

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 1000         | 127000  | 24.9509     | 542.686ms    |

| Done         | 509000  | 100         | 550.529ms    |

+--------------+---------+-------------+--------------+

Starting blockwise querying.

max rows per data block: 4348

number of reference data blocks: 4

number of query data blocks: 1

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 1000         | 119000  | 24.8954     | 483.231ms    |

| Done         | 478000  | 100         | 531.792ms    |

+--------------+---------+-------------+--------------+

Starting blockwise querying.

max rows per data block: 4348

number of reference data blocks: 4

number of query data blocks: 1

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 1000         | 127000  | 24.9509     | 547ms        |

| Done         | 509000  | 100         | 606.847ms    |

+--------------+---------+-------------+--------------+

Starting blockwise querying.

max rows per data block: 4348

number of reference data blocks: 4

number of query data blocks: 1

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 1000         | 127000  | 24.9509     | 472.077ms    |

| Done         | 509000  | 100         | 553.695ms    |

+--------------+---------+-------------+--------------+

Starting blockwise querying.

max rows per data block: 4348

number of reference data blocks: 4

number of query data blocks: 1

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 1000         | 127000  | 24.9509     | 492.672ms    |

| Done         | 509000  | 100         | 554.734ms    |

+--------------+---------+-------------+--------------+

Starting blockwise querying.

max rows per data block: 4348

number of reference data blocks: 4

number of query data blocks: 1

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 1000         | 119000  | 24.8954     | 484.793ms    |

| Done         | 478000  | 100         | 557.542ms    |

+--------------+---------+-------------+--------------+

Starting blockwise querying.

max rows per data block: 4348

number of reference data blocks: 4

number of query data blocks: 1

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 1000         | 127000  | 24.9509     | 555.598ms    |

| Done         | 509000  | 100         | 572.143ms    |

+--------------+---------+-------------+--------------+

Starting blockwise querying.

max rows per data block: 4348

number of reference data blocks: 4

number of query data blocks: 1

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 1000         | 128000  | 25.1473     | 510.863ms    |

| Done         | 509000  | 100         | 583.472ms    |

+--------------+---------+-------------+--------------+

Starting blockwise querying.

max rows per data block: 4348

number of reference data blocks: 4

number of query data blocks: 1

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 1000         | 127000  | 24.9509     | 540.602ms    |

| Done         | 509000  | 100         | 573.033ms    |

+--------------+---------+-------------+--------------+

Starting blockwise querying.

max rows per data block: 4348

number of reference data blocks: 4

number of query data blocks: 1

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 1000         | 120000  | 25.1046     | 489.1ms      |

| Done         | 478000  | 100         | 538.806ms    |

+--------------+---------+-------------+--------------+

Starting blockwise querying.

max rows per data block: 4348

number of reference data blocks: 4

number of query data blocks: 1

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 1000         | 127000  | 24.9509     | 490.566ms    |

| Done         | 509000  | 100         | 551.243ms    |

+--------------+---------+-------------+--------------+

### Create an SFrame with the distances from ‘dog’ test examples to the respective nearest neighbors in each class in the training data: The ‘distance’ column in dog_cat_neighbors above contains the distance between each ‘dog’ image in the test set and its nearest ‘cat’ image in the training set. The question we want to answer is how many of the test set ‘dog’ images are closer to a ‘dog’ in the training set than to a ‘cat’, ‘automobile’ or ‘bird’. So, next we will create an SFrame containing just these distances per data point. The goal is to create an SFrame called dog_distances with 4 columns:
* i. dog_distances[‘dog-dog’] ---- storing dog_dog_neighbors[‘distance’]
* ii. dog_distances[‘dog-cat’] ---- storing dog_cat_neighbors[‘distance’]
* iii. dog_distances[‘dog-automobile’] ---- storing dog_automobile_neighbors[‘distance’]
* iv. dog_distances[‘dog-bird’] ---- storing dog_bird_neighbors[‘distance’]
### Hint: You can create a new SFrame from the columns of other SFrames by creating a dictionary with the new columns, as shown in this example:
* new_sframe = graphlab.SFrame({'foo': other_sframe['foo'], 'bar': some_other_sframe['bar']})

In [77]:
dog_distances = graphlab.SFrame({'dog-dog': dog_dog_neighbors['distance'],
                                'dog-bird': dog_bird_neighbors['distance'],
                                'dog-cat': dog_cat_neighbors['distance'],
                                'dog-automobile': dog_automobile_neighbors['distance']})
cat_distances = graphlab.SFrame({'cat-cat': cat_cat_neighbors['distance'],
                                'cat-dog': cat_dog_neighbors['distance'],
                                'cat-bird': cat_bird_neighbors['distance'],
                                'cat-automobile': cat_automobile_neighbors['distance']})
bird_distances = graphlab.SFrame({'bird-bird': bird_bird_neighbors['distance'],
                                 'bird-cat': bird_cat_neighbors['distance'],
                                 'bird-dog': bird_dog_neighbors['distance'],
                                 'bird-automobile': bird_automobile_neighbors['distance']})
automobile_distances = graphlab.SFrame({'automobile-automobile': automobile_automobile_neighbors['distance'],
                                       'automobile-cat': automobile_cat_neighbors['distance'],
                                       'automobile-dog': automobile_dog_neighbors['distance'],
                                       'automobile-bird': automobile_bird_neighbors['distance']})

In [82]:
print dog_distances.head(1)
print cat_distances.head(1)
print bird_distances.head(1)
print automobile_distances.head(1)

+----------------+---------------+---------------+---------------+
| dog-automobile |    dog-bird   |    dog-cat    |    dog-dog    |
+----------------+---------------+---------------+---------------+
| 41.9579761457  | 41.7538647304 | 36.4196077068 | 33.4773590373 |
+----------------+---------------+---------------+---------------+
[1 rows x 4 columns]

+----------------+--------------+--------------+---------------+
| cat-automobile |   cat-bird   |   cat-cat    |    cat-dog    |
+----------------+--------------+--------------+---------------+
| 39.6710582792  | 38.074265869 | 34.623719208 | 37.4642628784 |
+----------------+--------------+--------------+---------------+
[1 rows x 4 columns]

+-----------------+---------------+---------------+---------------+
| bird-automobile |   bird-bird   |    bird-cat   |    bird-dog   |
+-----------------+---------------+---------------+---------------+
|  55.2246444674  | 51.3719536389 | 50.6626994751 | 50.5112353783 |
+-----------------+-----

### Computing the number of correct predictions using 1-nearest neighbors for the dog class: Now that you have created the SFrame dog_distances, you will learn to use the method
* .apply()
* on this SFrame to iterate line by line and compute the number of ‘dog’ test examples where the distance to the nearest ‘dog’ was lower than that to the other classes. You will do this in three steps:
#### i. Consider one row of the SFrame dog_distances. Let’s call this variable row. You can access each distance by calling, for example,
* row['dog-cat']
* which, in example table above, will have value equal to 36.4196077068 for the first row.
* Create a function starting with
* def is_dog_corrent(row):
* which returns 1 if the value for row[‘dog-dog’] is lower than that of the other columns, and 0 otherwise. That is, returns 1 if this row is correctly classified by 1-nearest neighbors, and 0 otherwise.
#### ii. Using the function is_dog_correct(row), you can check if 1 row is correctly classified. Now, you want to count how many rows are correctly classified. You could do a for loop iterating through each row and applying the function is_dog_correct(row). This method will be really slow, because the SFrame is not optimized for this type of operation.
* Instead, we will use the .apply() method to iterate the function is_dog_correct for each row of the SFrame.
* Read about using the .apply() method here.
* https://turi.com/learn/userguide/sframe/tabular-data.html
#### iii. Computing the number of correct predictions for ‘dog’: You can now call:
* dog_distances.apply(is_dog_correct)
* which will return an SArray (a column of data) with a 1 for every correct row and a 0 for every incorrect one. You can call:
    * .sum()
* on the result to get the total number of correctly classified ‘dog’ images in the test set!
* Hint: To make sure your code is working correctly, if you were to do the two steps above in this question to count the number of correctly classified ‘cat’ images in the test data, instead of ‘dog’, the result would be 548.

In [91]:
is_dog_correct = lambda row: (row['dog-dog'] < row['dog-cat']) and (row['dog-dog'] < row['dog-bird']) and (row['dog-dog'] < row['dog-automobile'])
is_cat_correct = lambda row: (row['cat-cat'] < row['cat-dog']) and (row['cat-cat'] < row['cat-bird']) and (row['cat-cat'] < row['cat-automobile'])
is_bird_correct = lambda row: (row['bird-bird'] < row['bird-cat']) and (row['bird-bird'] < row['bird-dog']) and (row['bird-bird'] < row['bird-automobile'])
is_automobile_correct = lambda row: (row['automobile-automobile'] < row['automobile-cat']) and (row['automobile-automobile'] < row['automobile-dog']) and (row['automobile-automobile'] < row['automobile-bird'])

In [95]:
dog_correct_count = dog_distances.apply(is_dog_correct).sum()
cat_correct_count = cat_distances.apply(is_cat_correct).sum()
bird_correct_count = bird_distances.apply(is_bird_correct).sum()
automobile_correct_count = automobile_distances.apply(is_automobile_correct).sum()

In [99]:
print "Dog Correct Predictions:", dog_correct_count
print "Cat Correct Predicitons:", cat_correct_count
print "Bird Correct Predictions:", bird_correct_count
print "Automobile Correct Predictions:", automobile_correct_count

Dog Correct Predictions: 678
Cat Correct Predicitons: 548
Bird Correct Predictions: 627
Automobile Correct Predictions: 947


####  Accuracy of predicting dog in the test data: Using the work you did in this question, what is the accuracy of the 1-nearest neighbor classifier at classifying ‘dog’ images from the test set? Save this result to answer the quiz at the end.

In [101]:
accuracy = float(dog_correct_count) / len(dog_image_test)
print accuracy

0.678
